### PR TITLE
Add support for x86 64 bit integer stack arguments.

### PR DESCRIPTION
**Pull request by andreistefanescu**

### CODE REVIEW

File: src/SAWScript/Crucible/LLVM/X86.hs [hunk 1]
```diff
@@ -1021,16 +1022,31 @@ allocateStack szInt balign = do
...
     Right sn -> pure sn
   fresh <- liftIO $ C.LLVM.LLVMPointer
     <$> W4.natLit sym 0
     <*> W4.freshConstant sym sn (W4.BaseBVRepr $ knownNat @64)
-  ptr <- liftIO $ C.LLVM.doPtrAddOffset bak mem' base =<< W4.bvLit sym knownNat (BV.mkBV knownNat szInt)
-  writeAlign <- integerToAlignment defaultStackBaseAlign
-  finalMem <- liftIO $ C.LLVM.doStore bak mem' ptr
+  rsp <- getReg Macaw.RSP regs
+  ptr <- liftIO $ C.LLVM.doPtrAddOffset bak mem rsp =<< W4.bvLit sym knownNat (BV.mkBV knownNat (-8))
+  let writeAlign = C.LLVM.noAlignment
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/LLVM/X86.hs:1048]: Why the change from `integerToAlignment defaultStackBaseAlign` to `noAlignment`?

File: src/SAWScript/Crucible/LLVM/X86.hs [hunk 2]
```diff
@@ -1152,36 +1168,57 @@ setArgs ::
...
-            _ -> fail "Argument does not fit into a single general-purpose register"
-      regs <- use x86Regs
-      newRegs <- liftIO . foldlM setRegSetupValue regs $ zip argRegs args
-      x86Regs .= newRegs
+        _ -> fail "Argument does not fit into a single general-purpose register"
+  regs <- use x86Regs
+  newRegs <- liftIO . foldlM setRegSetupValue regs $ zip argRegs args
+  x86Regs .= newRegs
+
+  let stackArgs = reverse $ Prelude.drop (length argRegs) args
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/LLVM/X86.hs:1201]: Writing each of the stack arguments in reverse order is somewhat peculiar—worth leaving a comment here saying why this is done.

File: intTests/test_llvm_x86_08/test.o
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply intTests/test_llvm_x86_08/test.o:1]: just the executable

File: intTests/test_llvm_x86_08/test.sh [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+set -e
+
+clang -c -emit-llvm -g -frecord-command-line test.c
+# clang -c -target x86_64 test.c
+# ld.lld -o test test.o
+clang -o test test.c
```
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply intTests/test_llvm_x86_08/test.sh:7]: done

File: src/SAWScript/Crucible/LLVM/X86.hs (hunk 2 above)
Comment (andreistefanescu) [reply src/SAWScript/Crucible/LLVM/X86.hs:1220]: done

File: intTests/test_llvm_x86_08/test.saw [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  llvm_execute_func
+    [ llvm_term {{ 0 : [64] }}
+    , llvm_term {{ 1 : [64] }}
+    , llvm_term {{ 2 : [64] }}
+    , llvm_term {{ 3 : [64] }}
+    , llvm_term {{ 4 : [64] }}
+    , llvm_term {{ 5 : [64] }}
+    , llvm_term x
+    , llvm_term {{ 1 : [64] }}
+    ];
```
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply intTests/test_llvm_x86_08/test.saw:18]: done

File: src/SAWScript/Crucible/LLVM/X86.hs (hunk 1 above)
Comment (andreistefanescu) [reply src/SAWScript/Crucible/LLVM/X86.hs:1064]: nice catch! Actually, this reveals a bug. The stack needs to be 16 byte aligned before the call instruction is executed, that is, before the return address has been pushed. The existing check is on `%rsp` after the return address has been pushed.

**Review by kquick — COMMENTED** [review]
Looks good, just a couple of minor questions, possibly deserving some inline comments.  Happy to approve when you are ready, just not doing so yet to avoid activating the mergebot.

File: src/SAWScript/Crucible/LLVM/X86.hs [hunk 5]
```diff
@@ -1152,37 +1184,62 @@ setArgs ::
...
+
+  -- x86-64 System V ABI specifies that: "Once registers are assigned, the
+  -- arguments passed in memory are pushed on the stack in reversed
+  -- (right-to-left21) order."
+  let stackArgs = reverse $ Prelude.drop (length argRegs) args
+  forM_ stackArgs $ \sval -> do
+    liftIO $ exceptToFail (typeOfSetupValue cc tyenv nameEnv sval) >>= \case
+      C.LLVM.PtrType _ -> pure ()
+      C.LLVM.IntType 64 -> pure ()
+      _ -> fail "Stack argument is not a 64 bit integer."
```
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply src/SAWScript/Crucible/LLVM/X86.hs:1225]: yes, this only handles 64-bit arguments. I didn't investigate what clang does, I *think* it doesn't promote it. The semantics if `fail`, since this can happen, and it's just not handled. Also, this doesn't handle `float`/`doable`, and other non-integer types.

File: src/SAWScript/Crucible/LLVM/X86.hs [hunk 6]
```diff
@@ -1152,37 +1184,62 @@ setArgs ::
...
+
+    regs' <- use x86Regs
+    rsp <- getReg Macaw.RSP regs'
+    rsp' <- liftIO $ C.LLVM.doPtrAddOffset bak mem rsp =<< W4.bvLit sym knownNat (BV.mkBV knownNat (-8))
```
[the remark this answers is not in the document]
Comment (andreistefanescu) [reply src/SAWScript/Crucible/LLVM/X86.hs:1229]: yes, it's just standard detection.

**Review by kquick — APPROVED** [review]
